Rewrite TensorImpl to use TensorTypeId.

ezyang opened the pull request:
Summary:
Translation to Backend happens immediately before we go into the
Type universe; otherwise we use TensorTypeId.

I allocated TensorTypeId corresponding exactly to existing ATen
Backend.  Only CPUTensorId and CUDATensorId are relevant in the
Caffe2 universe.

Differential Revision: D9184060

gchanan commented on aten/src/ATen/Formatting.cpp:
```diff
@@ -45,10 +45,6 @@ std::ostream& operator<<(std::ostream & out, Backend b) {
   return out << toString(b);
 }
 
-std::ostream& operator<<(std::ostream & out, ScalarType t) {
```
why this change?

ezyang replied:
The short answer is, good C++ practice is to put `operator<<` for user defined types in the same namespace as the type in question.

The long answer is, C++ has a feature called argument dependent lookup, whereby when it is looking for an implementation of `operator<<` to satisfy something like `std::cout << scalar_type`, it collects the namespaces of all of its arguments and looks for `operator<<` in those namespaces. This is why this program compiles: https://godbolt.org/g/a4ZQyK even though, ordinarily, `f::operator<<` would not be considered "in-scope" for a normal function call.

Now, in many cases, whether or not you put the `operator<<` definition in the global namespace or in your user namespace doesn't matter, because scope resolution will find the right one anyway. But if you have some `operator<<` definitions in a namespace, and you perform a `<<` in that namespace, the compiler will see that the namespaced `operator<<` is the relevant one, and will ignore your overloads in the global namespace. Example: https://godbolt.org/g/eZjXEs

You can "work around" this problem by explicitly telling the compiler to use the global `operator<<` with a using declaration https://godbolt.org/g/PjvoSR but there is a simpler, more excellent way to fix the problem: DON'T define `operator<<` in the global namespace; instead, rely on ADL to figure out what the correct namespace of `operator<<` should be.

(Also CC @cpuhrsch, this is why you should do it this way!)

gchanan commented on aten/src/ATen/SparseTensorImpl.cpp:
```diff
@@ -18,15 +30,13 @@ namespace at {
-    , values_(globalContext().getTypeOpt(toDense(backend), scalar_type)->tensor()) {
-      AT_ASSERT(backend == Backend::SparseCPU || backend == Backend::SparseCUDA);
-    }
+    , indices_(globalContext().getTypeOpt(sparseTensorIdToDenseBackend(type_id), ScalarType::Long)->tensor())
```
this will conflict with https://github.com/pytorch/pytorch/pull/9279 -- can you work out ordering with @yf225?

ezyang replied:
I'll do it after that PR lands.

gchanan commented on aten/src/ATen/core/TensorTypeId.cpp:
```diff
@@ -1,5 +1,9 @@
 #include "ATen/core/TensorTypeId.h"
 
+namespace at {
```
related to the other comment, why don't we do the opposite and move this to Formatting?  Is it because there are external usages that break?

ezyang replied:
I think it's silly to have to include `Formatting.h` if you want to `<<` types; the implementations are really simple, put them with the actual headers.

ezyang committed:
Back out "Back out "Move typeid.h to move to ATen/core""

fbshipit-source-id: 5960605c2abdf93e758cabb5605413457f96b251

ezyang committed:
Rewrite TensorImpl to use TensorTypeId. (#10278)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10278

Translation to Backend happens immediately before we go into the
Type universe; otherwise we use TensorTypeId.

I allocated TensorTypeId corresponding exactly to existing ATen
Backend.  Only CPUTensorId and CUDATensorId are relevant in the
Caffe2 universe.

Reviewed By: gchanan

Differential Revision: D9184060

fbshipit-source-id: ded568574c2db0c4503f0dc9d8c5e432608885f0
